### PR TITLE
BidirectionalGraph as subclass of DirectedAdjacencyGraph

### PR DESCRIPTION
**Pull request by jsjuni**

### CODE REVIEW

File: test/bidirectional_graph_test.rb [hunk 1]
```diff
@@ -0,0 +1,166 @@
+require 'test_helper'
+
+require 'rgl/adjacency'
+require 'rgl/bidirectional'
+
+include RGL
+include RGL::Edge
+
+class TestBidirectionalGraph < Test::Unit::TestCase
```
Comment (monora) [inline test/bidirectional_graph_test.rb:9]: We could in addition reuse `TestDirectedGraph` here only overriding setup method:

```suggestion
class TestDirectedGraphBidirected < TestDirectedGraph
  def setup
    @dg = BidirectionalAdjacencyGraph.new
    # TODO: refactor this in a protected setup function which we reuse here
    [[1, 2], [2, 3], [3, 2], [2, 4]].each do |(src, target)|
      @dg.add_edge(src, target)
    end
  end
end
```

All tests in the super test class should then pass.

**Review by monora — CHANGES_REQUESTED** [review]

File: lib/rgl/bidirectional.rb [hunk 2]
```diff
@@ -1,38 +1,128 @@
...
   # efficient access to in-edges typically requires more storage space,
   # and many algorithms do not require access to in-edges. For undirected
   # graphs, this is not an issue; because the in_edges() and out_edges()
   # functions are the same, they both return the edges incident to the vertex.
   #
-  module BidirectionalGraph
+  class BidirectionalGraph < DirectedAdjacencyGraph
+
+    OUT = 0
+    IN  = 1
```
[the remark this answers is not in the document]
Comment (jsjuni) [reply lib/rgl/bidirectional.rb:20]: That's essentially the first implementation I did. It was less intrusive than this one but also less efficient: The keys in the two vertex dicts are duplicated and all add/delete operations on vertices are duplicated. I'll make a PR for that version for comparison. It sounds closer to what you want.

File: lib/rgl/bidirectional.rb [hunk 3]
```diff
@@ -1,38 +1,128 @@
...
+  # The BidirectionalGraph concept refines DirectedAdjacencyGraph and adds the
   # requirement for efficient access to the in-edges of each vertex. This
-  # concept is separated from IncidenceGraph because, for directed graphs,
+  # concept is separated from DirectedAdjacencyGraph because, for directed graphs,
   # efficient access to in-edges typically requires more storage space,
   # and many algorithms do not require access to in-edges. For undirected
   # graphs, this is not an issue; because the in_edges() and out_edges()
   # functions are the same, they both return the edges incident to the vertex.
   #
-  module BidirectionalGraph
```
[the remark this answers is not in the document]
Comment (jsjuni) [reply lib/rgl/bidirectional.rb:15]: OK.

**Comment by monora** [comment]
Close and release in #83

**Pull request closed without merging**